web/api/v1/api.go

```diff
@@ -553,6 +601,57 @@ func (api *API) labelValues(r *http.Request) (result apiFuncResult) {
 	return apiFuncResult{vals, nil, warnings, closer}
 }
 
+// getLabelValuesFromSeriesSet extracts label values from matching series.
+func getLabelValuesFromSeriesSet(sets []storage.SeriesSet, name string) ([]string, storage.Warnings, error) {
+	set := storage.NewMergeSeriesSet(sets, storage.ChainedSeriesMerge)
+	labelValuesSet := make(map[string]struct{})
+	for set.Next() {
+		series := set.At()
+		labelValue := series.Labels().Get(name)
+		// Filter out empty value.
+		if labelValue == "" {
+			continue
+		}
+		labelValuesSet[labelValue] = struct{}{}
+	}
+
+	warnings := set.Warnings()
+	if set.Err() != nil {
+		return nil, warnings, set.Err()
+	}
+	// Convert the map to an array.
+	labelValues := make([]string, 0, len(labelValuesSet))
+	for key := range labelValuesSet {
+		labelValues = append(labelValues, key)
+	}
+	sort.Strings(labelValues)
+	return labelValues, warnings, nil
+}
+
+// getLabelNamesFromSeriesSet extracts label names from matching series.
+func getLabelNamesFromSeriesSet(sets []storage.SeriesSet) ([]string, storage.Warnings, error) {
```

Comment:
This should be inlined, I shouldn't have to jump around the file to find how this all works.

Reply:
Done.